src/main/scala/io/keen/client/scala/Client.scala

```diff
@@ -381,6 +405,166 @@ trait Writer extends AccessLevel {
+          val t: Thread = new Thread(r)
+          t.setDaemon(true)
+          t
+        }
+      }).asInstanceOf[ScheduledThreadPoolExecutor]
+
+      // schedule sending from our thread pool at a specific interval
+      tp.scheduleWithFixedDelay(new Runnable {
+        def run: Unit = {
+          try {
+            sendQueuedEvents
+          } 
+          catch {
+            case ex: Throwable => {
+              Logger.kerror("Failed to send queued events")
+              Logger.kerror(s"""$ex""")
+            }
+          }
+        }
+      }, 1, sendIntervalSeconds.toLong, TimeUnit.SECONDS)
+    }
+
+    tp
+
+  }
+
+  /**
+   * Sends all queued events, removing events from the queue as events are successfully sent.
+   */
+  def sendQueuedEvents(): Unit = synchronized {
+
+    val handleMap: TrieMap[String, ListBuffer[Long]] = eventStore.getHandles(projectId)
+    val handles: ListBuffer[Long] = ListBuffer.empty[Long]
+    val events: ListBuffer[String] = ListBuffer.empty[String]
+
+    // iterate over all of the event handles in the queue, by collection
+    for((collection, eventHandles) <- handleMap) {
+      // get each event, and its handle, then add it to a buffer so we can group the events 
+      // into smaller batches
+      for(handle <- eventHandles) {
+        handles += handle
+        events += eventStore.get(handle)
+      }
+
+      // group handles separately so we can use them to remove events from the queue once they've
+      // been succesfully added
+      val handleGroup: List[ListBuffer[Long]] = handles.grouped(batchSize).toList
+
+      // group the events by batch size, then publish them
+      for((batch, index) <- events.grouped(batchSize).zipWithIndex) {
+        // publish this batch
+        var response = Await.result(
+          addEvents(s"""{"$collection": [${batch.mkString(",")}]}"""),
+          DurationInt(batchTimeout).seconds
+        )
+
+        // handle addEvents responses properly
+        response.statusCode match {
+          case 200 | 201 => {
+            // log success
+            Logger.kinfo(s"""${response.statusCode} ${response.body} | Sent ${batch.size} queued events""")
+
+            // remove all of the handles for this batch
+            for(handle <- handleGroup(index)) {
+              eventStore.remove(handle)
+            }
+
+            // log removal
+            Logger.kinfo(s"""Removed ${handleGroup(index).size} events from the queue""")
+          }
+          case _ => {
+            // log but DO NOT remove events from queue
+            Logger.kerror(s"""${response.statusCode} ${response.body} | Failed to send ${batch.size} queued events""") 
+          }
+        }
+      }
+    }
+
+  }
+
+  /**
+   * Asynchronously sends all queued events.
+   */
+  def sendQueuedEventsAsync(): Unit = {
+
+    // use a thread pool for our async thread so we can use daemon threads
+    val tp = Executors.newSingleThreadExecutor(new ThreadFactory {
```

Comment:
Since you create this ThreadFactory twice I would just make it into its own class in a separate file.